tensorboard/summary_test.py

```diff
@@ -0,0 +1,81 @@
+# Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""API tests for the `tensorboard.summary` module.
+
+These tests are especially important because this module is the standard
+public entry point for end users, so we should be as careful as possible
+to ensure that we export the right things.
+"""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import collections
+
+import tensorflow as tf
+from tensorboard import summary
+
+
+STANDARD_PLUGINS = frozenset([
+    'audio',
+    'histogram',
+    'image',
+    'pr_curve',
+    'scalar',
+    'text',
+])
+
+# The subset of `STANDARD_PLUGINS` for which we do not currently have
+# functions to generate a summary protobuf outside of a TensorFlow
+# graph. This set should ideally be empty; any entries here should be
+# considered temporary.
+PLUGINS_WITHOUT_PB_FUNCTIONS = frozenset([
+    'pr_curve',  # TODO(@chihuahua): Fix this.
+])
+
+
+class SummaryExportsTest(tf.test.TestCase):
+
+  def test_each_plugin_has_an_export(self):
+    for plugin in STANDARD_PLUGINS:
+      self.assertIsInstance(getattr(summary, plugin), collections.Callable)
+
+  def test_plugins_export_pb_functions(self):
+    for plugin in STANDARD_PLUGINS:
```

Comment:
Our unit test best practices guide actually recommends against using things like for loops in unit tests. One notable exception is the Go language. But generally speaking, folks at Google err on the side of just having a lot of boilerplate to test each thing, so when the test runner tells you which method failed, you know the plugin for sure. You also wouldn't need to have special case logic for the pr_curve plugin.

But this is fine. I don't feel too strongly about it.

Reply:
I never really understood that logic. I agree that the burden of being "clearly correct" is stronger for test code than main code, but this loop is clearly correct. (It looks like we have 130 for-loops in test code currently.) (Plus, why should the Go language be any different?)

I thought about setting a custom message, but the case in which this will fail will almost always be the `getattr` failing, not that the user exports something that is not a function, and in that case the name of the plugin that failed will appear in the error message. (It's very frustrating that a custom message _replaces_ the default message instead of appending to it, or I'd use custom messages much more often.)